Mathlib/Data/Pi/Algebra.lean

```diff
@@ -0,0 +1,398 @@
+
+@[to_additive]
+theorem extend_div [Div γ] (f : α → β) (g₁ g₂ : α → γ) (e₁ e₂ : β → γ)
+    [∀ x, Decidable (∃ a, f a = x)] :
+    Function.extend f (g₁ / g₂) (e₁ / e₂) = Function.extend f g₁ e₁ / Function.extend f g₂ e₂ := by
+  funext x
+  simp [Function.extend_def, apply_dite₂]
+-- Porting note : This was the converted proof term not using tactic mode:
+-- `funext fun _ => by convert (apply_dite₂ (· / ·) _ _ _ _ _).symm`
+
+end Extend
+
+theorem surjective_pi_map {F : ∀ i, f i → g i} (hF : ∀ i, Surjective (F i)) :
+    Surjective fun x : ∀ i, f i => fun i => F i (x i) := fun y =>
+  ⟨fun i => (hF i (y i)).choose, funext fun i => (hF i (y i)).choose_spec⟩
+
+theorem injective_pi_map {F : ∀ i, f i → g i} (hF : ∀ i, Injective (F i)) :
+    Injective fun x : ∀ i, f i => fun i => F i (x i) :=
+  fun _ _ h => funext fun i => hF i <| (congr_fun h i : _)
+
+theorem bijective_pi_map {F : ∀ i, f i → g i} (hF : ∀ i, Bijective (F i)) :
+    Bijective fun x : ∀ i, f i => fun i => F i (x i) :=
+  ⟨injective_pi_map fun i => (hF i).injective, surjective_pi_map fun i => (hF i).surjective⟩
+
+
+#align bijective.injective Bijective.injective
+#align bijective.surjective Bijective.surjective
+
+end Function
+
+/-- If the one function is surjective, the codomain is trivial. -/
+@[to_additive "If the zero function is surjective, the codomain is trivial."]
+def uniqueOfSurjectiveOne (α : Type _) {β : Type _} [One β] (h : Function.Surjective (1 : α → β)) :
+    Unique β :=
+  h.uniqueOfSurjectiveConst α (1 : β)
+
+@[to_additive Subsingleton.pi_single_eq]
+theorem Subsingleton.pi_mul_single_eq {α : Type _} [DecidableEq I] [Subsingleton I] [One α]
+    (i : I) (x : α) : Pi.mulSingle i x = fun _ => x :=
+  funext fun j => by rw [Subsingleton.elim j i, Pi.mul_single_eq_same]
+
+namespace Sum
+
+variable (a a' : α → γ) (b b' : β → γ)
+
+@[simp, to_additive]
+theorem elim_one_one [One γ] : Sum.elim (1 : α → γ) (1 : β → γ) = 1 :=
+  Sum.elim_const_const 1
+
+@[simp, to_additive]
+theorem elim_mul_single_one [DecidableEq α] [DecidableEq β] [One γ] (i : α) (c : γ) :
+    Sum.elim (Pi.mulSingle i c) (1 : β → γ) = Pi.mulSingle (Sum.inl i) c := by
+  simp only [Pi.mulSingle, Sum.elim_update_left, elim_one_one]
+
+@[simp, to_additive]
+theorem elim_one_mul_single [DecidableEq α] [DecidableEq β] [One γ] (i : β) (c : γ) :
+    Sum.elim (1 : α → γ) (Pi.mulSingle i c) = Pi.mulSingle (Sum.inr i) c := by
+  simp only [Pi.mulSingle, Sum.elim_update_right, elim_one_one]
+
+@[to_additive]
+theorem elim_inv_inv [Inv γ] : Sum.elim a⁻¹ b⁻¹ = (Sum.elim a b)⁻¹ :=
+  (Sum.comp_elim Inv.inv a b).symm
+
+@[to_additive]
+theorem elim_mul_mul [Mul γ] : Sum.elim (a * a') (b * b') = Sum.elim a b * Sum.elim a' b' := by
+  ext x
+  cases x <;> rfl
+
+@[to_additive]
+theorem elim_div_div [Div γ] : Sum.elim (a / a') (b / b') = Sum.elim a b / Sum.elim a' b' := by
+  ext x
+  cases x <;> rfl
+
+end Sum
+
+-- Porting note : Atm it seems that to_additive does not copy the `instance` attribute
```

Comment:
Add a reference to #667?

Reply:
It seems that your comment did exactly that :)